### DISCUSSION TITLE
📢 PyScript NEXT (new core) Tech-Preview Release!! 📢

### DISCUSSION THREAD

**Discussion by fpliger** [pyscript/pyscript#1571] [Announcements]
📢  Announcement:  PyScript NEXT (new core) Tech-Preview Release!!  📢

After invaluable community feedback, many experiments in code, and careful cooperation with valued collaborators, we are thrilled to announce the tech preview of the new version of PyScript. This work marks a significant milestone in our journey towards PyScript's vision and what our community has been asking for.

Here are the highlights and key features of this groundbreaking technical Milestone, the Tech Preview Release:

**Massive Size Reduction**: We have optimized PyScript to offer a minimal download size. The core library is only ~10kb compressed and all the rest is downloaded only on demand. Combined with the very small footprint of MicroPython, you can have a fully working app in ~250kb, resulting in an astounding reduction of 98% compared to the previous version (~13.5Mb = ~11MB Pyodide + ~2.5MB Pyscript Classic). This makes PyScript more lightweight and faster to download with all the supported interpreters, ensuring a seamless user experience to the user.

**Versatile Execution Options:** With PyScript, you have the flexibility to choose the execution environment that best suits your needs. Whether you prefer executing your code on the main thread or leveraging the power of web workers for parallel processing, in a sync or async way, PyScript has got you covered.

**Blazing Fast MicroPython Interpreter:** The new PyScript core now incorporates a cutting-edge MicroPython interpreter, which delivers unparalleled speed and performance. Experience lightning-fast execution of your Python code in the browser environment.

**FFI support in MicroPython:** A new FFI API,  inspired by Pyodide's FFI has been added to Micropython and provides full access to Javascript and vice-versa. The goal of this API is to be as aligned as possible with Pyodide's FFI, to focus on the best experience possible to our users, considering the internal differences between these 2 interpreters.

**Experimental Interpreters:** In addition to the already exciting support for `Pyodide` and `Micropython` interpreters, as possible `Python` interpreters, this Tech Preview also adds support for 2 new experimental Interpreters: [wasmoon](https://github.com/ceifa/wasmoon), an interpreter that runs *Lua* on the browser and that, among the previous two interpreters, is fully compatible with all core features; and [ruby-wasm-wasi](https://github.com/ruby/ruby.wasm) an interpreter that adds *Ruby* to the list of programming languages currently supported :).

**Multiple Interpreter Support:** PyScript now supports the utilization of multiple interpreters simultaneously. This means you can take advantage of different languages (or Python versions) or configurations within a single application, opening up a world of possibilities for experimentation and optimization.

**Enhanced Compatibility:** Python on the browser is different from "traditional" Python. We have diligently worked on improving PyScript's compatibility with a wider range of Python code and libraries. Specifically, execution on Web Workers should allow us to support blocking calls and other aspects not available until now. This also includes the possibility of better debugging capabilities. We now think we have the foundation upon which to build these features in the future.

We invite you to give the tech preview a spin, build something amazing and share your feedback with us. Your participation in the project is at the heart of what we do, and essential for guaranteeing PyScript allows you to build the amazing things you want and fulfills its potential for everyone.

To get started, visit the [Tech Preview showcase](https://pyscript.net/tech-preview/pyscript.core/), where you can find examples that you can clone and play with and check out the [related documentation](https://github.com/pyscript/pyscript/blob/next/pyscript.core/docs/README.md)!

Have fun and build cool stuff.

Happy coding!

**NOTE:** This Tech Preview is by no means a replacement for the current PyScript official project and it's not meant to be. It's an early release of the many new possibilities that are coming next in official PyScript. The current APIs provided by the Tech Preview are likely going to change and align closely with the APIs of official PyScript. DO NOT USE THIS TECH PREVIEW IN PRODUCTION.

The PyScript Team

**Comment by ajasja** [comment]
Amazing! How is such a reduction in size possible? Wow.

**Reply by WebReflection** [reply]
`@pyscript/core` is minimalistic on purpose, enabling different interpreters as its main goal, augmenting their scope only with the new *XWorker* related feature and not much else. As *MicroPython* entered the game, and it offers `import js` and other *DOM* interaction goodness *Pyodide* offered before, we can now bootstrap a *Python-like* "_env_" in about 250K (I think that's gzipped or brotlied size) in a snap, still providing full *pyodide* features out of the box via a different script, or a *worker*, that won't ever block the main page/thread/UI while operating. So that's it: the entry point can be super fast and super small, considering how much *Python* can run via *MicroPython*, and the rest can be more complex on demand, without issues for the main page itself. I hope this made some sense, happy to explain details further, if needed!

**Reply by ajasja** [reply]
Thank you all clear! So basically, MicroPython is replaced by Pyodide. This is probably good enough for a lot of webpages that only print stuff. And then if needed the full pyodide can be loaded in a separate thread.

**Reply by fpliger** [reply]
Yes, the footprint of the core itself is orders of magnitude smaller. In addition to that, MicroPython is also orders of magnitude smaller than Pyodide. Now, these 2 aren't exactly fully comparable (in the sense that MicroPython is not a full implementation of full CPython. 

Now, after this proof of concept made true by the Tech Preview, the work is not done. We'll be working on integrating it into the PyScript Platform itself and adding more and more higher-level features offered by the current PyScript classic (like, the `repl` or Plugins) so that the new version will be powered by the new core. One of the goal is to keep the base PyScript framework as small as possible and add features dynamically only if users are using it. (Take the `repl` for instance, it's very big compared to the rest and if an user is not using it, they shouldn't be paying that price)

**Comment by THEGOLDENPRO** [comment]
Omg, this is such amazing news to hear. **98% reduction**, that is going to completely flip the tables. My number one concern about PyScript had always been the load times, I can't wait to try out the new core. ❤️ Thank you, the PyScript team.

**Comment by sskagemo** [comment]
I agree, it is amazing to have Micropython almost instantly! Really impressive work!

Have I understood it correctly that currently there is no "out of the box" REPL in the PyScript Next?

I have tried to make it work and read the documentation, so I just would like to have it confirmed before I spend more time on it. I am planning to set up a blog (in Norwegian) and thought it would be great to have the code-examples be runnable through pyscript, without the download/start-up time.

**Reply by WebReflection** [reply]
> Have I understood it correctly that currently there is no "out of the box" REPL in the PyScript Next?

the *PyScript Next* is on the making, as *custom type* (`<script type="py">`) but be assured we've planned already to make it available ASAP ... maybe the TP, even if updated, won't have the same exact *REPL* we offered before, but that is, I suppose, the goal of a TP, keep listening to feedback and bring in everything desirable out of the box.

Apologies if this is not the answer you were looking for, but "*classic PyScript*" is still up and running.

P.S. as multiple interpreters entered the game, of course the *REPL* might be also slightly different across them, and MicroPython in here is no exception, simply because it's not exactly 1:1 to Python (Pyodide) so we need to release carefully what can work, hence what won't, across interpreters.

**Reply by sskagemo** [reply]
Great, thank you! I'll definitly keep watching this space :-)

**Comment by prasad-shembekar** [comment]
🚀 Thank you for the amazing PyScript NEXT announcement! Excited to explore the new features and provide feedback. Keep up the fantastic work! 👏🎉
